proposals/003-installer-guidelines.md

```diff
@@ -0,0 +1,102 @@
+## Introduction
+
+The Haskell Foundation would like see a central Haskell download page in which the Haskell curious
+could easily download and run a installer to provide a standard Haskell environment that supports
+the principal build tools (`cabal-install` and `stack`) on the main platforms (Linux, macOS and
+Windows). All Haskell introductory material (text books, blog posts, etc.) should be able to point
+to the Haskell Foundation download page and work on the assumption that it will yield their
+widely-supported Haskell development environment.
+
+To this end this proposal covers two things:
+
+  * To publish a set of minimal requirements for a Haskell installer that we are prepared to
+    recommend along with a roadmap of the minimal requirements that we are looking to add in the
+    future. (The initial requirements will not require precompiled libraries or editor/IDE
+    environments but we would like to move in that direction, for example.)
+
+  * To provide a prototype Windows installer for cabal-install-based Haskell development
+    environments based on the stack codebase. This is the coverage we are missing at the moment and
+    it is easy to do. We anticipate it becoming quickly obsoleted as the more stablished installers
+    cover this space, but by providing this we immediately guarantee its coverage.
+
+  * To provide a simplified Haskell download page that prominently displays installers that
+    meet our guidelines.
+
+On the basis of early discussions with the principal stakeholders we believe that it is possible to
+quickly agree an initial set of requirements and build a Windows installer.
+
+
+## Background and Motivation
+
+The exact configuration of Haskell development environments has recently been a contentious issue
+leading to regrettable fragmentation and confusion. It should be a priority of the Haskell
+Foundation to lead the way in restoring harmony for the health of the Haskell community and its
+ecosystem.  The Haskell Foundation should also be well placed to do this with its broad
+representation at board level and many sponsors and friends in the community.
```

Comment:
Strong 👎 

This is very opinionated. Fragmentation happens for various reasons. Some of them may be good reasons. This proposal should solely focus on improving windows haskell user experience and not start debates about cabal vs stack or make attempts to explain why confusion exists.

It is clear that windows support needs work and more options on windows are welcome. That's all we need to do here.

Reply:
I strongly disagree with you here.  The path may be longer than I'd like to get there amicably, but having competing installers to get started with Haskell is clearly a bad user experience, and frankly sort of an embarrassing state of affairs.  I'd love for people to agree on a set of standards mainly so that we can get to the point where one installer can meet those standards and become the clear choice.

Reply:
> but having competing installers to get started with Haskell is clearly a bad user experience

Is it though? Competition in open source is natural. A "getting started" guide doesn't have to expose the user to all choices. I've said that numerous times.

> I'd love for people to agree on a set of standards mainly so that we can get to the point where one installer can meet those standards and become the clear choice.

This is clearly out of scope of this PR.

And I believe the HF has a lot of work to do before it should consider driving forward or moderating such an effort.

Reply:
This is why it's clear to me that all sorts of concerns are mixed up and mushed into one project:

1. Downloads page
2. UX for haskell newcomers 
3. Installer that works across all platforms
4. Smooth integration of standalone installers with stack
5. Unification of tools or implementations

All of these are connected, but different things and need their own consideration.

Reply:
There's a clear story that connects the points you mention.  2, 3, and 4 are requirements.  The downloads page is involved because it's the motivation behind the requirements: anyone is free to create any half-functional Haskell installer they like as a personal project, but if they want it linked from the Haskell downloads page, then it should meet community standards.

Unification is, indeed, not part of this proposal.  We can disagree on whether that's a mistake, but in any case setting a high standard for all tools is a step toward eventual unification.  My understanding is that stack maintains installer tools not because they want to, but because they saw certain needs that weren't met by general Haskell install tools (for one: Windows installs, and installing stack in the first place, which are now resolved; but also: allowing installs of GHC to be triggered by the stack build tool.)  If these needs were met by a tool like ghcup, maybe we could get out of this embarrassing situation in the first place and have a clean story for how to install Haskell.  I definitely look forward to this happening, and see this proposal as a stepping stone to get there even though this proposal doesn't ask for unification.

I cannot think of any other major programming language whose install page says "here are several competing ways to install".  That's the embarrassing situation we're in here.  When we tell someone how to install Haskell, there ought to be an answer of record to that question.  Note that this doesn't mean there can't be other options, such as installing from an Ubuntu PPR, or Chocolatey for Windows, or some custom setup script if Haskell is bundled into a special-use project as I already do with CodeWorld.  And if someone creates a better general-purpose install tool, that answer of record can change.  But right now we're paralyzing newcomers with a choice they shouldn't have to make.

Reply:
> There's a clear story that connects the points you mention. 2, 3, and 4 are requirements.

2 isn't really specified properly in this proposal. 3 has been done since ~2 months. 4 is stuck, because stack developers haven't responded to my issues or PRs yet and I don't blame them after all this confusion/drama.

> The downloads page is involved because it's the motivation behind the requirements: anyone is free to create any half-functional Haskell installer they like as a personal project, but if they want it linked from the Haskell downloads page, then it should meet community standards.

And it's unclear what "downloads page" that is.

>  I definitely look forward to this happening, and see this proposal as a stepping stone to get there even though this proposal doesn't ask for unification.

So far, HFs involvement in these matters have caused more damage and disconnection than there was before. So again: I'm not sure HF is the right place to drive forward the unification vision. The maintainers of the tools have to find a point of minimal consensus and work towards something. This is probably not the right time to think about it.

Unification is less a technical challenge than it is a collaboration challenge. Stack maintainers need control over the installation procedure to ensure their UX philosophy doesn't get compromised. GHCup has a very radical unix philosophy. Can we unify those? Probably. But when you depend on another tool/library, you're also relying on their maintainers. Now you have increased round-trip time for issues, patches, etc. because you need to communicate and can't just push fixes. Now you have to possibly debate features and so on and so on.

The easiest way forward, imo, is to first provide *GHC installation user hooks* for stack, which I have already implemented. That way users get full control over how stack installs GHC, be it ghcup, nix or their own shell script code.

Everything else is reaching for the stars and won't get us anywhere atm.

> I cannot think of any other major programming language whose install page says "here are several competing ways to install". That's the embarrassing situation we're in here. When we tell someone how to install Haskell, there ought to be an answer of record to that question. Note that this doesn't mean there can't be other options, such as installing from an Ubuntu PPR, or Chocolatey for Windows, or some custom setup script if Haskell is bundled into a special-use project as I already do with CodeWorld. And if someone creates a better general-purpose install tool, that answer of record can change. But right now we're paralyzing newcomers with a choice they shouldn't have to make.

I can't answer this question in regards to the proposal, since I really don't know what HF wants, what download page they envision etc. But I'm working hard towards improving the haskell.org side of this matter.

And I agree: the primary download page should NOT give choice, but make a choice. This proposal says the opposite, with the motivation to please all political parties. Except, new users don't care about that.

Reply:
> I really don't know what HF wants

As a collective body, HF is unable to desire anything on its own accord, unless enshrined in meeting minutes.  Tech Track meetings are open for members of public, and I encourage everyone interested to [attend](https://calendar.google.com/calendar/u/0/embed?src=g5pug6rj631f31d4ief6avdodc@group.calendar.google.com). Minutes of previous Tech Track meetings are available [here](https://drive.google.com/drive/folders/17M5yX8nFNSINdty_l0NlmwYWaLTuu7iI?ths=true).

Reply:
> And I agree: the primary download page should NOT give choice, but make a choice. This proposal says the opposite, with the motivation to please all political parties. Except, new users don't care about that.

Okay, I was under the impression from some previous conversations with you that you did *not* agree with that.  I'm relieved to hear we are on the same page here.  I think it would be great if the Haskell Foundation proposal here would reflect that agreement.  My question (for others, not for you) is: Are there parties involved who don't agree with that ultimate goal?  Or is this something we all want and just have disagreements over how to get there?  It's okay if we don't get there in a single step!

It seems like one way to get there would be to define an ambitious line to aim for (one that represents our authentic combined goals as a community), and concede that there will still be this not-great situation with multiple installers in the short run, but that eventually someone will hopefully cross that line, and the community can unify around that answer.  That might not be the right path... who knows... but that's how I interpreted this proposal, as trying to draw that line.  I don't think it's a perfect line yet, and I've left my comments as to why.  But I don't think it's a non-productive conversation to have.

I'd particularly hate to see the haskell.org committee go one way, and the Haskell Foundation a different way.  That would be a real blow to the community, IMO.  If we're at risk of that happening, then priority number one (before passing any proposal like this one) should be to figure out why, and how we can build some consensus.

Reply:
> Okay, I was under the impression from some previous conversations with you that you did not agree with that.

I was fine with this not being pursued for the sake of getting an "Installer guidelines" proposal through that tries to stay neutral. This was my collaboration effort, but I did voice that I think it's not a good idea. I do understand the concern to stay neutral. It's a different approach. It can be reasonable for other reasons. My point was simply that end users don't care.

>  My question (for others, not for you) is: Are there parties involved who don't agree with that ultimate goal?

We're not at the point where the involved parties have an open conversation. And so my suggestion is to postpone these discussions and not force them.

> I'd particularly hate to see the haskell.org committee go one way, and the Haskell Foundation a different way.

I think haskell.org and HF are very different organizations with different approaches. The [HF page](https://haskell.foundation/en/affiliates) says haskell.org committee is in the process of affiliation and judging from certain public comments it seems there is some effort to synchronize, but they're ultimately separate entities and can come to separate conclusions. I don't see any problem with that.

Reply:
Sorry for the delay in response.  My point about the two orgs going different directions is this: to the best of my knowledge, there is not a single person who thinks a beneficial outcome from this process would be for new Haskell install instructions to be posted to haskell.foundation, while the existing instructions remain at haskell.org.  That would clearly be terrible.

So if this proposal might be interpreted to say that, then it should be fixed to explicitly say that we're talking about the install instructions on haskell.org.  If it cannot be fixed, or if agreement cannot be found between the haskell.org committee and the Haskell Foundation, then the proposal should be rejected, since accepting it under those circumstances would be worse than doing nothing.

This isn't meant to challenge the autonomy of either organization.  It's just an observation that in this _specific_ situation, the HF adopting a proposal at odds with haskell.org would be a disaster.